contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -0,0 +1,743 @@
+    `containerType` of `""`. Since this does not match `"DEBUG"` the special
+    handling of Debug Containers is backwards compatible.
+
+### Security Considerations
+
+Debug Containers have no additional privileges above what is available to any
+`v1.Container`. It's the equivalent of configuring an shell container in a pod
+spec but created on demand.
+
+Admission plugins that guard `/exec` must be updated for the new parameters. In
+particular, they should enforce the same container image policy on the `Image`
+parameter as is enforced for regular containers. During the alpha phase we will
+additionally support a container image whitelist as a kubelet flag to allow
+cluster administrators to easily constraint debug container images.
+
+### Additional Consideration
+
+1.  Debug Containers are intended for interactive use and always have TTY and
+    Stdin enabled.
+1.  There are no guaranteed resources for ad-hoc troubleshooting. If
+    troubleshooting causes a pod to exceed its resource limit it may be evicted.
+1.  There's an output stream race inherent to creating then attaching a
+    container which causes output generated between the start and attach to go
+    to the log rather than the client. This is not specific to Debug Containers
+    and exists because Kubernetes has no mechanism to attach a container prior
+    to starting it. This larger issue will not be addressed by Debug Containers,
+    but Debug Containers would benefit from future improvements or work arounds.
+1.  We do not want to describe Debug Containers using `v1.Container`. This is to
+    reinforce that Debug Containers are not general purpose containers by
+    limiting their configurability. Debug Containers should not be used to build
+    services.
+1.  Debug Containers are of limited usefulness without a shared PID namespace.
+    If a pod is configured with isolated PID namespaces, the Debug Container
+    will join the PID namespace of the target container. Debug Containers will
+    not be available with runtimes that do not implement PID namespace sharing
+    in some form.
+
+## Implementation Plan
+
+### Alpha Release
+
+#### Goals and Non-Goals for Alpha Release
+
+We're targeting an alpha release in Kubernetes 1.9 that includes the following
+basic functionality:
+
+*   Support in the kubelet for creating debug containers in a running pod
+*   A `kubectl debug` command to initiate a debug container
+*   `kubectl describe pod` will list status of debug containers running in a pod
+
+Functionality will be hidden behind an alpha feature flag and disabled by
+default. The following are explicitly out of scope for the 1.9 alpha release:
+
+*   Exited Debug Containers will be garbage collected as regular containers and
+    may disappear from the list of Debug Container Statuses.
+*   Security Context for the Debug Container is not configurable. It will always
+    be run with `CAP_SYS_PTRACE` and `CAP_SYS_ADMIN`.
```

Comment:
So only a cluster admin should ever use debug containers?

Reply:
Also, pods don't have security context set in many cases.  Exec implicitly escalating to root is bad.

Reply:
This should be configurable at debug time, but that depends on the API. Some of my proposed API changes addressed this, but SIG Node and the API reviewers deadlocked on which was best. Our compromise was to proceed to alpha with the minimum possible API change, which doesn't include a configurable security context.

Created https://github.com/kubernetes/kubernetes/issues/53188 to track this.

/cc @thockin

Reply:
> Our compromise was to proceed to alpha with the minimum possible API change, which doesn't include a configurable security context.

you have a pointer to where that discussion happened? no one from the auth/psp side was involved afaik

Reply:
I don't understand, you've been involved since the first draft of the proposal 10 months ago, and your input has always been most welcome. It's not too late, what would you like to see changed?

Reply:
> It's not too late, what would you like to see changed?

there has to be control over the security aspects (uid/gid, capabilities, pid/network/ipc namespace, etc) of containers that are launched on the nodes. Omitting that control ("Security Context for the Debug Container is not configurable"), and defaulting to privileged levels ("It will always be run with `CAP_SYS_PTRACE` and `CAP_SYS_ADMIN`.") is a particular problem.

Reply:
Agreed, and I created an issue to track adding a configurable security context. Until then this feature is an alpha feature requiring cluster admins to enable it.

I don't think it's a good idea to encode an entire securityContext as an HTTP query parameter. Are there any fields you can't live without during the alpha? What about a toggle for SYS_ADMIN?

Reply:
> I created an issue to track adding a configurable security context. Until then this feature is an alpha feature requiring cluster admins to enable it.

When there are known gaps required for exiting alpha status that have yet-undetermined implications on the API for this feature, it doesn't seem wise to start making changes to stable APIs in master.

> I don't think it's a good idea to encode an entire securityContext as an HTTP query parameter. Are there any fields you can't live without during the alpha? What about a toggle for SYS_ADMIN?

I agree that encoding is ugly, and I don't a great idea there, but I think it's a mistake to not deal with this in terms of a standard `Container` definition:
* many parts of the security context has utility in some debug scenario:
  * running many debug tools require elevated capabilities
  * other scenarios would want to run as a specific uid/gid (plenty of overlap with https://github.com/kubernetes/kubernetes/issues/30656)
* admission plugins that make decisions on pods do so in terms of PodSpec and `Container` structs. What is the benefit of coming up with a different structure to describe the same thing?

Reply:
Jordan,

Is your concern mainly that the structure is not a v1.Container ?  Or is the concern around the (ab)use of /exec?  I am not super keen on the /exec bits, but I want this to unstick and forward progress feels like a better tradeoff than another LITERAL year of wheel-spinning.

If we work through the changes required to plumb security down, even though not fully implementing for alpha, does that satisfy?  If not, can you explain exactly what you'd like to see for forward motion?

Reply:
> Is your concern mainly that the structure is not a v1.Container ? Or is the concern around the (ab)use of /exec? 

Mostly the former (the latter is ugly, but not fatal). I have three main concerns:

1. This proposes to build the container spec implicitly in the kubelet, given an image. That means admission doesn't have the information needed to allow/deny/modify that container spec, and if the kubelet changes how it defines the container spec release-to-release, admission can't even know what the kubelet will do.
2. There's not a clear path forward to support more aspects of the container spec (control over pull policy, uid/gid, capabilities, env, etc, all seem likely and reasonable)
3. The proposed defaults don't match the defaults for Container specs (specifically around security settings / capabilities), which would make transitioning to dealing with this as a `v1.Container` difficult

If we don't deal with this as a `v1.Container` consistently through the stack (api/admission/kubelet), we're making every layer have to understand two ways containers can be expressed, which seems like a mistake.

Reply:
@verb how bad would it be to pass a full v1.Container?

Reply:
@thockin Not bad, It's what the kubelet does internally and I've had it working in a prototype.

We would use the API described in [Alternative 1](https://github.com/kubernetes/community/pull/649/files#diff-5cfb31b40ca47511743d0545d5697aa0R144) to POST a v1.Container (wrapped in a new top level object) to a new /debug subresource. Then the client would perform a separate /attach.

The API reviewers had concerns about this being a novel use of the API. Nothing else POSTs to a subresource.

Both the API reviewers and SIG Node reviewers had concerns about using v1.Container being confusing or communicating the wrong intent to the user. Debug Containers are not general purpose containers and should not be used to build services or for routine operations. Some prefer to consider "extended exec" rather than "configurable container".

Most of the fields of v1.Container do not apply to Debug Containers and should be rejected if configured (lifecycle, livenessProbe, ports, readinessProbe, resources, stdin, stdinOnce, terminationMessagePath, terminationMessagePolicy, tty, volumeMounts). We can do this with a validation whitelist, though it would be simpler to pass a securityContext rather than a full v1.Container.

Reply:
> The API reviewers had concerns about this being a novel use of the API. Nothing else POSTs to a subresource.

I thought scheduler POSTs to pod/binding subresource, clients POST to pod/eviction subresource, etc. Or is there a distinction between POST and PUT here? (Which I can never keep straight.)

Reply:
@davidopp Oh, that's good news then. When I prototyped this ~6 months ago I recall needing a couple of changes in the apiserver in order to make it work, but maybe those were specific to upgrading a connection to streaming after a POST. It's been a long ride.

Let's figure out a way to move this forward. Since we already had agreement among the reviewers at the time, and now we want to renegotiate that agreement based on new reviewers, I suggest that take the form of a new PR to amend the proposal where the new and old reviewers can work out their conflicting requirements. I'll prepare a diff.

Reply:
>> The API reviewers had concerns about this being a novel use of the API. Nothing else POSTs to a subresource.
>
> I thought scheduler POSTs to pod/binding subresource, clients POST to pod/eviction subresource, etc. Or is there a distinction between POST and PUT here? (Which I can never keep straight.)

Not sure about subresources, but there are issues with both PUT and POST as part of *websocket* requests (not all clients support them)